wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/CheEnvironmentValidator.java

```diff
@@ -33,92 +45,269 @@
-            validateMachine(machineCfg, envName);
+        checkArgument(env.getMachines() != null && !env.getMachines().isEmpty(),
+                      "Environment '%s' doesn't contain machine with 'ws-agent' agent",
+                      envName);
+
+        List<String> missingServices = env.getMachines()
+                                          .keySet()
+                                          .stream()
+                                          .filter(machineName -> !composeEnvironment.getServices()
+                                                                                    .containsKey(machineName))
+                                          .collect(toList());
+        checkArgument(missingServices.isEmpty(),
+                      "Environment '%s' contains machines that are missing in environment recipe: %s",
+                      envName, Joiner.on(", ").join(missingServices));
+
+        List<String> devMachines = env.getMachines()
+                                      .entrySet()
+                                      .stream()
+                                      .filter(entry -> entry.getValue()
+                                                            .getAgents()
+                                                            .contains("ws-agent"))
+                                      .map(Map.Entry::getKey)
+                                      .collect(toList());
+
+        checkArgument(devMachines.size() == 1,
+                      "Environment '%s' should contain exactly 1 machine with ws-agent, but contains '%d'. " +
+                      "All machines with this agent: %s",
+                      envName, devMachines.size(), Joiner.on(", ").join(devMachines));
+
+        // needed to validate different kinds of dependencies in services to other services
+        Set<String> servicesNames = composeEnvironment.getServices().keySet();
+
+        composeEnvironment.getServices()
+                          .forEach((serviceName, service) -> validateMachine(serviceName,
+                                                                             env.getMachines().get(serviceName),
+                                                                             service,
+                                                                             envName,
+                                                                             servicesNames));
+
+        // check that order can be resolved
+        try {
+            startStrategy.order(composeEnvironment);
+        } catch (IllegalArgumentException e) {
+            throw new IllegalArgumentException(
+                    format("Compose services order of environment '%s' is not resolvable. Error: %s",
+                           envName, e.getLocalizedMessage()));
         }
     }
 
-    private void validateMachine(MachineConfig machineCfg, String envName) throws IllegalArgumentException {
+    protected void validateMachine(String machineName,
```

Comment:
Why not to use native foreach?, it looks better for me + i think it will work faster and will use less memory as you don't need to create stream for each collection you'r checking